### Practicing with QISKit

In this notebook I'm experiencing basic QISKit functionalities. I'm going to create a 3-qbit Bell State, and ensure the qbits are actualy intricated.

In [3]:
from qiskit import QuantumProgram, QISKitError

In [6]:
# Creating a QuantumProgram instance
Q_program = QuantumProgram()

In [18]:
# Performing 3-qbits intrication using a local simulator

try:
    # Creating a 3-qbit quantum register
    qr = Q_program.create_quantum_register("qr", 3)
    # Creating a 3-bits classical register
    cr = Q_program.create_classical_register("cr", 3)
    # Creating a quantum circuit that involves the previously created registers
    qc = Q_program.create_circuit("intrication", [qr], [cr])
    
    # Hadamard gate on the first qbit to get some superposition
    qc.h(qr[0])
    # CNOT gate to intricate the first and second qbits
    qc.cx(qr[0], qr[1])
    # CNOT gate to intricate the second and third qbits
    qc.cx(qr[1], qr[2])
    
    # Performs the measures
    qc.measure(qr, cr)
    
    # Compile and run the simulation
    result = Q_program.execute(["intrication"], backend="local_qasm_simulator",
                               shots=1024)
    
    # Show the results
    print(result)
    print(result.get_data("intrication"))
    
except QISKitError as ex:
    print("There was an error : {}".format(ex))

COMPLETED
{'counts': {'000': 489, '111': 535}}


##### Results

The simulation yielded the following results :

  |000> : 489 over 1024 shots
  
  |111> : 535 over 1024 shots
 

Yay, the qbits are either all 0s, or all 1s, they are fully intricated ! But that was only a simulation assuming perfect computation, not an actual quantum computer. Let's move on to the real deal.

In [25]:
import Qconfig

Q_program.set_api(Qconfig.APItoken, Qconfig.config["url"], verify=False,
                  hub=Qconfig.config["hub"],
                  group=Qconfig.config["group"],
                  project=Qconfig.config["project"])

q_result = Q_program.execute(["intrication"], backend="ibmqx2", shots=1024)

print(q_result)
print(q_result.get_data("intrication"))

-- Ignoring SSL errors.  This is not recommended --
COMPLETED
{'time': 22.839900970458984, 'counts': {'00001': 36, '00110': 43, '00000': 451, '00100': 9, '00011': 42, '00010': 7, '00101': 8, '00111': 428}, 'date': '2018-01-03T18:33:13.975Z'}


##### Results

   |000> : 451
   
   |001> : 36
   
   |010> : 7
   
   |011> : 42
   
   |100> : 9
   
   |101> : 8
   
   |110> : 43
   
   |111> : 428
   
   (1024 shots)
   
We can see that the implementation is imperfect, thus yielding other results that |000> and |111>. But the vast majority of them are what was expected : |000> and |111>.